x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStopTrainedModelDeploymentAction.java

```diff
@@ -137,22 +167,26 @@ private void normalUndeploy(Task task, PersistentTasksCustomMetadata.PersistentT
                 }
             }
         );
-
         super.doExecute(task, request, finalListener);
     }
 
-    void waitForTaskRemoved(Set<String> taskIds, StopTrainedModelDeploymentAction.Request request,
+    void waitForTaskRemoved(String modelId,
+                            TrainedModelAllocation trainedModelAllocation,
+                            StopTrainedModelDeploymentAction.Request request,
                             StopTrainedModelDeploymentAction.Response response,
                             ActionListener<StopTrainedModelDeploymentAction.Response> listener) {
-        persistentTasksService.waitForPersistentTasksCondition(persistentTasks ->
-                persistentTasks.findTasks(MlTasks.TRAINED_MODEL_DEPLOYMENT_TASK_NAME, t -> taskIds.contains(t.getId())).isEmpty(),
-            request.getTimeout(), ActionListener.wrap(
-                booleanResponse -> {
-                    listener.onResponse(response);
-                },
+        final Set<String> nodesOfConcern = trainedModelAllocation.getNodeRoutingTable().keySet();
+        client.admin()
```

Comment:
I think calling this directly here requires the caller of the stop API to have the appropriate privileges. The way around this is to call the `ListTasksAction` with ML origin.

Reply:
> I think calling this directly here requires the caller of the stop API to have the appropriate privileges.

I will switch to using the `ML_ORIGIN` for sure.